django_postmark_utils/admin.py

```diff
@@ -1,3 +1,271 @@
-from django.contrib import admin
+from email.utils import formatdate
+import logging
+import pickle
+import traceback
 
-# Register your models here.
+from django.conf import settings
+from django.contrib import admin, messages
+from django.core.mail import get_connection
+from django.core.mail.message import make_msgid
+from django.core.mail.utils import DNS_NAME
+from django.urls import reverse
+from django.utils.html import format_html
+from django.utils.translation import ugettext_lazy as _
+
+from . import app_settings
+from .models import Bounce, Delivery, Email, Message
+from .utils import ResendEmailMessage
+
+
+logger = logging.getLogger(__name__)
+
+
+class ReadOnlyModelAdminMixin(object):
+
+    def has_add_permission(self, request):
+        return False
+
+    def has_delete_permission(self, request, obj=None):
+        return False
+
+    def get_actions(self, request):
+        actions = super().get_actions(request)
+        del actions['delete_selected']
+        return actions
+
+
+class EmailInline(ReadOnlyModelAdminMixin, admin.TabularInline):
+
+    model = Email
+    fields = (
+        'is_resend',
+        'msg_id',
+        'date',
+        'subject',
+        'sending_error',
+        'delivery_msg_id',
+        'delivery_error_code',
+        'num_of_bounces',
+        'num_of_deliveries',
+    )
+    readonly_fields = (
+        'is_resend',
+        'msg_id',
+        'date',
+        'subject',
+        'sending_error',
+        'delivery_msg_id',
+        'delivery_error_code',
+        'num_of_bounces',
+        'num_of_deliveries',
+    )
+
+    def num_of_bounces(self, obj):
+        return obj.bounces.count()
+    num_of_bounces.short_description = _("bounces")
+
+    def num_of_deliveries(self, obj):
+        return obj.deliveries.count()
+    num_of_deliveries.short_description = _("deliveries")
+
+
+@admin.register(Message)
+class MessageAdmin(ReadOnlyModelAdminMixin, admin.ModelAdmin):
```

Comment:
We need something in the message list display to show that a message has a problem. Perhaps a counts of all associated bounces and deliveries?

Reply:
This messages list display will be the main mechanism for looking through and finding failed messages. We need to be able to see subject, from and to, on this view and ideally to be able to search over them. It might mean storing these values on the message object instead of the email object?